Add pluggable device profiler support (#389)
tensorflow/community: rfcs/20210513-pluggable-profiler-for-tensorflow.md

2. Core TensorFlow populates `TF_ProfilerRegistrationParams` and passes it in a call to `TF_InitProfiler`. Core TensorFlow owns the memory of `TF_ProfilerRegistrationParams`'s `profiler` and `profiler_fns` struct.
    * In `TF_InitProfiler`, the plugin populates `TF_ProfilerRegistrationParams`'s `profiler` and `profiler_fns`.
3. Core TensorFlow registers a profiler creation function to ProfilerFactory based on `profiler` and `profiler_fns` populated by the plugin during the initialization time.
4. Core Tensorflow will create a `PluggableProfiler` during [ProfilerSession](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/lib/profiler_session.cc#L109) setup.

#### Protobuf class
Profiler uses `XSpace` to store the performance data collected by backends. With this type of data structure, TensorFlow's profiler tools can generate various views of performance, such as timeline, memory consumption, performance of every TensorFlow op and set of summaries. `XSpace` is C++ object generated by protobuf toolchain with a predefined structure in [xplane.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/protobuf/xplane.proto#L9).

> On the plugin side, plugin will collect performance data with its own profiler infrastructures. When TensorFlow invokes `CollectData()`, the plugin transforms the performance data to `XSpace` and then serializes `XSpace` to the buffer provided by TensorFlow. To successfully serialize the object, plugin writers should keep a copy of `xspace.proto`, and make it exactly the same as that in the TensorFlow side.

Reviewer: ```suggestion
On the plugin side, plugin will collect performance data with its own profiler infrastructures. When TensorFlow invokes `CollectData()`, the plugin transforms the performance data to `XSpace` and then serializes `XSpace` to the buffer provided by TensorFlow. To successfully serialize the object, plugin writers should keep a copy of `xspace.proto`, and make it exactly the same as that on the TensorFlow side.
```

```c++
#define TP_MAJOR 0
#define TP_MINOR 0
#define TP_PATCH 1

#ifdef __cplusplus
extern "C" {
#endif

typedef struct TP_Profiler {
  size_t struct_size;
  void* ext; // free-form data set by plugin.
  const char* type;
} TP_Profiler;

#define TP_PROFILER_STRUCT_SIZE TF_OFFSET_OF_END(TP_Profiler, type)

typedef struct TP_ProfilerFns {
  size_t struct_size;

  void* ext; // reserved for future use.
  // Starts profiling.
  void (*start)(const TP_Profiler* profiler, TF_Status* status);
  // Stops profiling.
  void (*stop)(const TP_Profiler* profiler, TF_Status* status);

  // Saves collected profile data into XSpace and serializes it to the buffer.
  // If this have been called, subsequent calls might
  // return empty data.
  void (*collect_data_xspace)(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status);
} TP_ProfilerFns;

#define TP_PROFILER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(TP_ProfilerFns, collect_data_xspace)

typedef struct TF_ProfilerRegistrationParams {
  size_t struct_size;
  void* ext; // reserved for future use

  // TensorFlow Profiler C API version.
  int32_t major_version;
  int32_t minor_version;
  int32_t patch_version;

>   [in/out] Memory owned by core but attributes within are populated by the plugin.

Reviewer 2: Needs to be a comment: //

Author: Done

  TP_Profiler* profiler;

>   [in/out] Memory owned by core but attributes within are populated by the plugin.

Reviewer 2: Needs to be a comment: //

Author: Done

#ifdef __cplusplus
} // extern "C"
#endif
```

### Usage Example
This section provides some pseudo code to show what core TensorFlow and plugin's code may look like.

#### Core TensorFlow
* **ProfileOptions support**

  To enable [ProfileOptions](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/profiler_options.proto#L6) support for `PluggableProfiler`, we add a new device type `PLUGGABLE_DEVICE` in the `enum DeviceType` field.
  ```c++
  // Next ID: 11
  message ProfileOptions {
    // Some default values of option are not proto3 default values. Use this version
    // to determine if we should use default option value instead of proto3
    // default value.
    uint32 version = 5;
  
    enum DeviceType {
      UNSPECIFIED = 0;
      CPU = 1;
      GPU = 2;
      TPU = 3;
      PLUGGABLE_DEVICE = 4;
    }
      // Device type to profile/trace: (version >= 1)
    // DeviceType::UNSPECIFIED: All registered device profiler will be enabled.
    // DeviceType::CPU: only CPU will be profiled.
    // DeviceType::GPU: only CPU/GPU will be profiled.
    // DeviceType::TPU: only CPU/TPU will be profiled.

>     // DeviceType::PLUGGABLE_DEVICE: all pluggable devices with profilers enabled will be profiled. 

Reviewer 2: only CPU and pluggable devices with profilers enabled will be profiled

Author: Done

      // Get size of buffer required for Plugin to serialize XSpace ito.
      size_t size_in_bytes;
      profiler_fns_->collect_data_xspace(profiler_, /*buffer=*/nullptr,
         				&size_in_bytes, c_status.get());
  
      // Prepare an appropriately sized buffer.
      if (size_in_bytes > 0) {
        std::vector<uint8_t> buffer(size_in_bytes);
        profiler_fns_->collect_data_xspace(profiler_, buffer.data(),
         				  &size_in_bytes, c_status.get());
        // Deserialize XSpace from the buffer and return it.
        XSpace plugin_space;
        plugin_space.ParseFromArray(buffer.data(), buffer.size());
        for (XPlane& plugin_plane: *plugin_space.mutable_planes()) {
          XPlane* plane = space->add_planes();
          plane->Swap(&plugin_plane);
        }
      }
      return tensorflow::StatusFromTF_Status(c_status.get());
    }
  
   private:
    PluggableProfiler(TP_ProfilerFns* profiler_fns, TP_Profiler* profiler) 
      : profiler_fns_(profiler_fns), profiler_(profiler) {}
    
    TP_ProfilerFns* profiler_fns_;
    TP_Profiler* profiler_;
  }
  ```

* **PluggableProfiler Initialization and Registration**

>   Core TensorFlow will load `TF_InitProfiler` from plugin's dynamic library installed under "…python_dir.../site-packages/tensorflow-plugins" and pass the address of `TF_InitProfiler` symbol to `InitPluginProfiler` to do initialization and registration. TensorFlow retrieves `TF_ProfilerRegistrationParams` from the plugin and does the compatibility checks. If passed, TensorFlow will register the PluggableProfiler creation function to the [ProfilerFactroy](https://github.com/tensorflow/tensorflow/blob/aa855d21ac89e6649ec782ef4efd4e126b37d79d/tensorflow/core/profiler/lib/profiler_factory.cc#L39).

Reviewer: ```suggestion
  Core TensorFlow will load `TF_InitProfiler` from plugin's dynamic library installed under "…python_dir.../site-packages/tensorflow-plugins" and pass the address of `TF_InitProfiler` symbol to `InitPluginProfiler` to do initialization and registration. TensorFlow retrieves `TF_ProfilerRegistrationParams` from the plugin and does the compatibility checks. If passed, TensorFlow will register the PluggableProfiler creation function to the [ProfilerFactory](https://github.com/tensorflow/tensorflow/blob/aa855d21ac89e6649ec782ef4efd4e126b37d79d/tensorflow/core/profiler/lib/profiler_factory.cc#L39).
```

  ```c++
  class PluggableProfilerFactory{
   public:
    PluggableProfilerFactory(TP_Profiler profiler,
  			   void (*destroy_profiler)(TP_Profiler*),
  			   TP_ProfilerFns profiler_fns,
  			   void (*destroy_profiler_fns)(TP_ProfilerFns*))
    	: profiler_(std::move(profiler)),
            destroy_profiler_(destroy_profiler),
            profiler_fns_(std::move(profiler_fns)),
            destroy_profiler_fns_(destroy_profiler_fns) {}
  
    ~PluggableProfilerFactory() {
      destroy_profiler_(&profiler_);

>       destroy_profiler_fns_(&profiler_fns_);

Reviewer 2: Needs closing brace: }

Author: Done

    std::unique_ptr<tensorflow::profiler::ProfilerInterface> CreatePluggableProfiler(
        const ProfileOptions& options) {
      return PluggableProfiler::CreatePluggableProfiler(options, &profiler_,
                                                      &profiler_fns_);
    }

   private:
      TP_Profiler profiler_{TP_PROFILER_STRUCT_SIZE};
      void (*destroy_profiler_)(TP_Profiler*);
      TP_ProfilerFns profiler_fns_{TP_PROFILER_FNS_STRUCT_SIZE};
      void (*destroy_profiler_fns_)(TP_ProfilerFns*);
    }
  
  }
  
  Status InitPluginProfiler(TFInitProfilerFn init_fn) {
    TF_ProfilerRegistrationParams params{TF_PROFILER_REGISTRATION_PARAMS_STRUCT_SIZE};
    TP_Profiler profiler{TP_PROFILER_STRUCT_SIZE};
    TP_ProfilerFns profiler_fns{TP_PROFILER_FNS_STRUCT_SIZE};
    params.major_version = TP_MAJOR;
    params.minor_version = TP_MINOR;
    params.patch_version = TP_PATCH;
    params.profiler = &profiler;
    params.profiler_fns = &profiler_fns;

>     OwnedTFStatus c_status(TF_NewStatus());

Reviewer 2: std::unique_ptr<TF_Status, TFStatusDeleter> c_status(TF_NewStatus());

for consistency with the rest of the code

Author: Done

    init_fn(&params, c_status.get());
    TF_RETURN_IF_ERROR(tensorflow::StatusFromTF_Status(c_status.get()));
    TF_RETURN_IF_ERROR(ValidateTPProfilerRegistrationParams(params));
    TF_RETURN_IF_ERROR(ValidateTPProfiler(profiler));
    TF_RETURN_IF_ERROR(ValidateTPProfilerFns(profiler_fns));
  
    PluggableProfilerFactory factory(std::move(profiler),
  				   params.destroy_profiler,
  				   std::move(profiler_fns),
  				   params.destroy_profiler_fns);

>     )

Reviewer 2: Orphan closing parenthesis, remove.

Author: Done

  
    tensorflow::profiler::ProfilerFactory::RegisterProfilerFactory(
      [factory = std::move(factory)](const ProfileOptions& options) {

>         factory.CreatePluggableProfiler(options);

Reviewer 2: missing 'return'

Author: Done

  params->destroy_profiler_fns = profiler_destroy_profiler_fns;
}
```

## Requirements for the Profiler Library
* If there's no actual device to be profiled (e.g., if running on a machine w/o the required device), the profiler does nothing on Start/Stop and produces no output on CollectData.
* If the profiled program does not use the device, the profiler produces no output on CollectData.
e.g., returns 0 as the required buffer size for the serialized XSpace.
* Restartability: It must be possible to Start/Stop the profiler many times. (Since a single TP_Profiler object is created per library).
* No (perceivable) overhead: Profiling should not change the observed performance of the profiled job.
e.g. the step time of a TF training job should be similar with profiling on or off.
* No OOM: Profiling should not require significant additional memory resources.
Also, profiling should not cause the process to be killed due to running out-of-memory (OOM).
* No leaks: any resources (memory) acquired for handling a profiling request should be released by the end of the request.
Repeated profiling requests should not increase resource utilization over time.

> * No memory corruption: profiling should not corrupt memory due to dangling pointers.

Reviewer: ```suggestion
* No memory corruption: profiling should not corrupt memory. e.g. due to dangling pointers.
```

> # Modular TensorFlow profiler C API

Reviewer: @deven-amd
Please have a look at this RFC. I think it would be cleaner to have all the GPU profilers consolidated behind a common interface.

Reviewer 3: Hi @yisitu, I am working closely with Deven in AMD. Please also CC me for further discussions for the TF profiler. Thanks.

Reviewer: Sure!

## **User Benefit**

This RFC provides a plugin infrastructure for extending third-party device profilers.

## **Design Proposal**

### Design Overview

This RFC is intended to provide a set of C APIs for plugin writers to implement and register their own pluggable profilers. To make C APIs portable, we propose serialized `XSpace` and `RunMetadata` as the objects to pass between TensorFlow framework and plugin. When the framework invokes `CollectData()`, the plugin serializes `XSpace` and `RunMetadata` into sufficiently sized buffers provided by the framework. Subsequently, the framework deserializes these buffers back into `XSpace` and `RunMetadata`, and generates a trace view and a set of summaries based on these collected data.

- Xspace:
<div align=center>
<img src=20210513-pluggable-profiler-for-tensorflow/Xspace.png>
</div>

- RunMetadata:
<div align=center>
<img src=20210513-pluggable-profiler-for-tensorflow/RunMetadata.png>
</div>

> To achieve the goal, this RFC extends the TensorFlow profiler class hierarchy to add a new profiler named `PluggableProfiler` which is built on top of a set of C APIs, all plugin writers who want to integrate their own device profilers to  current TensorFlow runtime only need to implement Profiler C APIs(shown as diagram of Architecture overview).

Author: done

> To achieve the goal, this RFC extends the TensorFlow profiler class hierarchy to add a new profiler named `PluggableProfiler` which is built on top of a set of C APIs, all plugin writers who want to integrate their own device profilers to  current TensorFlow runtime only need to implement Profiler C APIs(shown as diagram of Architecture overview).

Reviewer: Thank you.

<div align=center>
<img src=20210513-pluggable-profiler-for-tensorflow/Architecture.png>
</div>

### Versioning Strategy and Stability
* **Profiler C API**
    Version strategy of Profiler C API follows Semantic Versioning 2.0.0 ([semver](http://semver.org/)). Each release version has a format `MAJOR.MINOR.PATCH`, as outlined in [TensorFlow version compatibility](https://www.tensorflow.org/guide/versions#semantic_versioning_20). Struct size is used to track compatibility. More details can be found in [StreamExecutor C API Versioning Strategy RFC](https://github.com/tensorflow/community/blob/master/rfcs/20200612-stream-executor-c-api/C_API_versioning_strategy.md)
* **XPlane and RunMetadata**
   The compatibility of `XPlane` and `RunMetadata` between plugin and TensorFlow follows the same compatibility [rules](https://developers.google.com/protocol-buffers/docs/cpptutorial?hl=en#extending-a-protocol-buffer) and [guarantees](https://developers.google.com/protocol-buffers/docs/proto3?hl=en#updating) as protobuf library.

> 

Author: Done

#### Protobuf class
Profiler uses `RunMetadata` and `XSpace` to store the performance data collected by backends. With these two data, TensorFlow's profiler tools can generate various views of performance, such as timeline, memory consumption, performance of every TensorFlow op and set of summaries. `RunMedata` and `XSpace` are C++ objects generated by protobuf toolchain with a predefined structure in [config.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/config.proto#L721) and [xplane.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/protobuf/xplane.proto#L9).
On the plugin side, plugin will collect performance data with its own profiler infrastructures. When TensorFlow invokes `CollectData()`, plugin transforms the performance data to `RunMetadata` and `XSpace` and then serializes `RunMetadata` and `XSpace` to the buffer provided by TensorFlow. To successfully serialize the object, plugin writers should keep a copy of `config.proto` and `xspace.proto`, and make it exactly the same as that in the TensorFlow side.

> ### Detailed API

Author: done,

```c++
#define TP_MAJOR 0
#define TP_MINOR 0
#define TP_PATCH 1

#ifdef __cplusplus
extern "C" {
#endif

typedef struct TP_Profiler {
  size_t struct_size;
  void* ext; // free-form data set by plugin.
  const char* type;
} TP_Profiler;

#define TP_PROFILER_STRUCT_SIZE TF_OFFSET_OF_END(TP_Profiler, type)

typedef struct TP_ProfilerFns {
  size_t struct_size;

  void* ext; // reserved for future use.

> 
>   void (*start)(const TP_Profiler* profiler, TF_Status* status); // Starts profiling

Reviewer: ```suggestion
  // Starts profiling
  void (*start)(const TP_Profiler* profiler, TF_Status* status);
```

  void (*stop)(const TP_Profiler* profiler, TF_Status* status); // Stops profiling.

  // Saves collected profile data into XSpace and serializes it to the buffer.
  // After this or the overload above are called once, subsequent calls might
  // return empty data.
  void (*collect_data_run_metadata)(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status)

  // Saves collected profile data into run_metadata and serializes it to the buffer.
  // After this or the overload below are called once, subsequent calls might
  // return empty data.
  void (*collect_data_xspace)(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status);

> };

Reviewer: ```suggestion
} TP_ProfilerFns;
```

> #define TF_PROFILER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(TP_ProfilerFns, collect_data_xspace)

Author: done

> #define TF_PROFILER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(TP_ProfilerFns, collect_data_xspace)

Author: done

typedef struct TF_ProfilerRegistrationParams {
  size_t struct_size;
  void* ext; // reserved for future use

  // TensorFlow Profiler C API version.
  int32_t major_version;
  int32_t minor_version;
  int32_t patch_version;

>   TP_Profiler* profiler; // output, set by plugin
>   TP_ProfilerFns* profiler_fns; // output set by plugin
>   // Clean up fields inside TP_Profiler that were allocated
>   // by the plugin. `profiler` itself should not be deleted here.
>   void (*destroy_profiler)(TP_Profiler* profiler); // out, set by plugin
>   void (*destroy_profiler_fns)(TP_ProfilerFns* profiler_fns);

Reviewer: ```suggestion
  [in/out] Memory owned by core but attributes within are populated by the plugin.
  TP_Profiler* profiler;
  [in/out] Memory owned by core but attributes within are populated by the plugin.
  TP_ProfilerFns* profiler_fns;
  // [out] Pointer to plugin's `TP_Profiler` clean up function. 
  // Cleans up fields inside `TP_Profiler` that were allocated
  // by the plugin. `profiler` itself must not be deleted by the plugin.
  void (*destroy_profiler)(TP_Profiler* profiler);
  // [out] Pointer to plugin's `TP_ProfilerFns` clean up function. 
  // Cleans up fields inside `TP_ProfilerFns` that were allocated
  // by the plugin. `profiler_fns` itself must not be deleted by the plugin.
  void (*destroy_profiler_fns)(TP_ProfilerFns* profiler_fns);
```

  ```c++
  // Next ID: 11
  message ProfileOptions {
    // Some default value of option are not proto3 default value. Use this version
    // to determine if we should use default option value instead of proto3
    // default value.
    uint32 version = 5;
  
    enum DeviceType {
      UNSPECIFIED = 0;
      CPU = 1;
      GPU = 2;
      TPU = 3;
      PLUGGABLE_DEVICE = 4;
    }
      // Device type to profile/trace: (version >= 1)
    // DeviceType::UNSPECIFIED: All registered device profiler will be enabled.
    // DeviceType::CPU: only CPU will be profiled.
    // DeviceType::GPU: only CPU/GPU will be profiled.
    // DeviceType::TPU: only CPU/TPU will be profiled.
    // DeviceType::PLUGGABLE_DEVICE: all pluggable devices with profiler enabled will be profiled. 
    DeviceType device_type = 6;
  ```
  Due to `device_type` here is enum, we can't differentiate between multiple pluggable profilers, so we define a common device type `PLUGGABLE_DEVICE` for them, if `ProfileOptions` is configured with `PLUGGABLE_DEVICE` type, then all the registered pluggable profilers will be enabled.

* **Plugin Profiler Initialization**

>   Core TensorFlow will load `TF_InitProfiler` from plugin's dynamic library installed under "…python_dir.../site-packages/tensorflow-plugins" and pass the handle to `InitPluginProfiler` to do initialization. Here we define a `PluginInterfaceFatory` to store all the registered pluggable profilers, all of pluggable profilers in this factory will be enabled if `device_type` in `ProfileOptions` is configured as `PLUGGABLE_DEVICE`.

Author: done, TFInitProfilerFn is expected to define in profiler_internal.h, which is not a public API

          plane->Swap(&plugin_plane);
        }
      }
      Status s = tensorflow::StatusFromTF_Status(c_status.get());
      return s;
    }

   private:
     TP_Profiler profiler_;
     void (*destroy_profiler_)(TP_Profiler*);
     TP_ProfilerFns profiler_fns_;
     void (*destroy_profiler_fns_)(TP_ProfilerFns*);
  };

  ```
* **PluggableTracer Registration**

  `PluggableTracer` is the implementation of [ProfilerInterface](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/lib/profiler_interface.h#L33), it controls all the registered pluggable profilers. For example, if `device_type` of `ProfileOptions` is set as `PLUGGABLE_DEVICE`, all the registered pluggable profilers will be enabled, when `ProfilerSession` invokes `profiler->Start()`, all the registered pluggable profilers will start the profiling.
  ```c++
  class PluggableTracer : public ProfilerInterface {
   public:
     explicit PluggableTracer(std::vector<PluginTracerInterface> plugin_interfaces)
             : plugin_interfaces_(plugin_interfaces) {}

     ~PluggableTracer() override {}

     Status Start() override {
       for (auto& profiler_interface : plugin_interfaces_) {

>          profiler_interface.DoStart().IgnoreError();

Reviewer: (Here and below) Errors should not be ignored, and instead should be Update() and returned.

https://abseil.io/docs/cpp/guides/status#keeping-track-of-the-first-error-encountered

Author: Thanks for the comments, the consideration here is whether one plugin profiler failed will affect other plugin profiler, I see ProfilerSession, it seems it ignored all the profiler status. 
```
 61 
 62 Status ProfilerSession::CollectData(profiler::XSpace* space) {
 63   mutex_lock l(mutex_);
 64   TF_RETURN_IF_ERROR(status_);
 65 #if !defined(IS_MOBILE_PLATFORM)
 66   LOG(INFO) << "Profiler session collecting data.";
 67   for (auto& profiler : profilers_) {
 68     profiler->Stop().IgnoreError();
 69   }
 70 
 71   for (auto& profiler : profilers_) {
 72     profiler->CollectData(space).IgnoreError();
 73   }
 74 
 75   if (active_) {
 76     // Allow another session to start.
 77     profiler::ReleaseProfilerLock();
 78     active_ = false;
 79   }
 80 
 81   PostProcessSingleHostXSpace(space, start_time_ns_);
 82 #endif
 83 
 84   return Status::OK();
 85 }
```

Reviewer: That is a bug in ProfilerSession. The correct behavior should be to `Update()` status for each profiler and not return early upon failure. This way we track the first error encountered and resume calling to other plugin profilers.

Author: Done

        options.device_type() != ProfileOptions::UNSPECIFIED) {
      return nullptr;
    }
    return absl::make_unique<PluggableTracer>(*PluginInterfaceFactory::GetPluginTracerInterfaces());
  }

  auto register_pluggable_tracer_factory = [] {
    RegisterProfilerFactory(&CreatePluggableTracer);
    return 0;
  }();

  ```

#### Plugin
Define functions that implement `Start()`, `Stop()`, `CollectData()`:
```c++
void profiler_start(const TP_Profiler* profiler, TF_Status* status) {
  // Enable profiler
  ...
}

void profiler_stop(const TP_Profiler* profiler, TF_Status* status) {
  // Disable Profiler
  ...
}

void profiler_collect_data_run_metadata(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status) {
  RunMetadata metadata = get_my_run_metadata(); // Plugin generates RunMetadata based on collected profiler data.
  *size_in_bytes = metadata.ByteSizeLong(); // get the size of RunMetadata
  if (buffer == nullptr) {
    return; // TensorFlow will first get the size of RunMetadata, then allocate the big enough buffer and pass it to plugin for retrieving RunMetadata.
  }
  metadata.SerializeToArray(buffer, metadata.ByteSizeLong());
}

> void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status) {
>   Xspace xspace = get_my_xspace(); // Plugin generates Xspace based on collected profiler data.
>   *size_in_bytes = xspace.ByteSizeLong(); // get the size of Xspace
>   if (buffer == nullptr) {
>     return; // TensorFlow will first get the size of Xspace, then allocate the big enough buffer and pass it to plugin for retrieving Xspace.
>   }
>   xspace.SerializeToArray(buffer, xspace.ByteSizeLong());
> }

Reviewer: 1) Need to check the return value of `SerializeToArray()`
2) Need to pass the user's buffer size to SerializeToArray().
3) Need to set the out parameter `status`.

```suggestion
void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status) {
  Xspace xspace = get_my_xspace(); // Plugin generates Xspace based on collected profiler data.
  size_t buffer_size_in_bytes = * size_in_bytes;
  *size_in_bytes = xspace.ByteSizeLong(); // get the size of Xspace
  if (buffer == nullptr) {
    return; // TensorFlow will first get the size of Xspace, then allocate the big enough buffer and pass it to plugin for retrieving Xspace.
  }
  bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
 // TODO: set status to FAILED_PRECONDITION if success is false, OK otherwise.
}
```

# Modular TensorFlow profiler C API
| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [389](https://github.com/tensorflow/community/pull/389)|
| **Author(s)** | Zhoulong Jiang (zhoulong.jiang@intel.com), Yiqiang Li (yiqiang.li@intel.com), Eric Lin (eric.lin@intel.com), Jianhui Li (jian.hui.li@intel.com) |
| **Sponsor**   | Situ Yi (yisitu@google.com)                          |
| **Updated**   | 2021-05-13                                          |

## **Objective**

Provide general profiler C API to allow new device profilers to modularly connect to the current TensorFlow runtime.

This RFC is based on the Modular TensorFlow  [RFC](https://github.com/tensorflow/community/pull/77), which aims at extending the TensorFlow design to plug in capabilities like adding a new device profiler.

## **Motivation**

> Performance is a key consideration of successful ML research and production solutions. TensorFlow profiler provides a set of good tools to help user better understand the hardware resource consumption(time and memory) of various TensorFlow operations(ops) as well as perforamnce bottlenecks of TensorFlow models. [PluggableDevice](https://github.com/tensorflow/community/pull/262) provides a mechanism which allows third-party devices to be modularly integrated to current TensorFlow runtime. However, current TensorFlow does not provide a stable application binary infterface(ABI) for registering a new pluggable device specific profiler in a modular way. We propose a C API wrapper of methods in [ProfilerInterface](https://github.com/tensorflow/tensorflow/blob/0a3773ed7b4c1fc60486dddfacbe9a5cbf2b2bdd/tensorflow/core/profiler/lib/profiler_interface.h#L33) as an ABI-stable way to register a custom profiler. The pluggable profiler discovery and initialization are transparent to end users. As long as the profiler plugin libraries follow the design described in this RFC, it can be plugged to TensorFlow proper and register a new profiler into ProfilerFactory.

Reviewer: ```suggestion
Performance is a key consideration of successful ML research and production solutions. TensorFlow profiler provides a set of good tools to help user better understand the hardware resource consumption(time and memory) of various TensorFlow operations(ops) as well as performance bottlenecks of TensorFlow models. [PluggableDevice](https://github.com/tensorflow/community/pull/262) provides a mechanism which allows third-party devices to be modularly integrated to current TensorFlow runtime. However, current TensorFlow does not provide a stable application binary interface(ABI) for registering a new pluggable device specific profiler in a modular way. We propose a C API wrapper of methods in [ProfilerInterface](https://github.com/tensorflow/tensorflow/blob/0a3773ed7b4c1fc60486dddfacbe9a5cbf2b2bdd/tensorflow/core/profiler/lib/profiler_interface.h#L33) as an ABI-stable way to register a custom profiler. The pluggable profiler discovery and initialization are transparent to end users. As long as the profiler plugin libraries follow the design described in this RFC, it can be plugged to TensorFlow proper and register a new profiler into ProfilerFactory.
```
